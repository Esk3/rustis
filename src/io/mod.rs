use std::io::{BufReader, BufWriter};

use thiserror::Error;

use crate::resp::Value;

pub trait Io {
    fn read_value(&mut self) -> IoResult<Value>;
    fn write_value(&mut self, value: Value) -> IoResult<usize>;
}

pub trait ValueIo {}
pub trait MessageIo {}

#[derive(Error, Debug)]
pub enum IoError {
    #[error("end of input")]
    EndOfInput,
    #[error("io error {0}")]
    Io(#[from] std::io::Error),
}

pub type IoResult<T> = Result<T, IoError>;

#[derive(Debug)]
pub struct TcpStream<'a> {
    r: BufReader<&'a std::net::TcpStream>,
    w: BufWriter<&'a std::net::TcpStream>,
}

impl<'a> TcpStream<'a> {
    #[must_use]
    pub fn new(stream: &'a std::net::TcpStream) -> Self {
        Self {
            r: BufReader::new(stream),
            w: BufWriter::new(stream),
        }
    }
}
impl Io for TcpStream<'_> {
    fn read_value(&mut self) -> IoResult<Value> {
        todo!("read value from stream")
    }

    fn write_value(&mut self, value: Value) -> IoResult<usize> {
        todo!("write {value:?} to stream")
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum NetworkMessage {
    Input(Input),
    Output(Output),
}

#[derive(Debug, PartialEq, Eq)]
pub enum Input {
    Ping,

    Get(String),
    Set {
        key: String,
        value: String,
        expiry: Option<std::time::Duration>,
        get: bool,
    },

    Multi,
    CommitMulti,

    ReplConf(ReplConf),
    Psync,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Output {
    Pong,
    Get(Option<String>),
    Set,

    Multi,
    Queued,

    ReplConf(ReplConf),
    Psync,
    Null,
    Ok,
    Array(Vec<Self>),
}

#[derive(Debug, PartialEq, Eq)]
pub enum ReplConf {
    ListingPort(u16),
    Capa(String),
    GetAck(i32),
    Ack(i32),
}

#[cfg(test)]
pub mod tests {
    use crate::resp::Value;

    use super::{Io, IoResult};

    #[derive(Debug)]
    pub struct MockIo {
        inputs: Vec<Value>,
        outputs: Vec<Value>,
    }

    impl MockIo {
        pub fn new<I, O>(inputs: I, outputs: O) -> Self
        where
            I: IntoIterator<Item = Value>,
            <I as std::iter::IntoIterator>::IntoIter: std::iter::DoubleEndedIterator,
            O: IntoIterator<Item = Value>,
            <O as std::iter::IntoIterator>::IntoIter: std::iter::DoubleEndedIterator,
        {
            Self {
                inputs: inputs.into_iter().rev().collect(),
                outputs: outputs.into_iter().rev().collect(),
            }
        }
    }
    impl Io for MockIo {
        fn read_value(&mut self) -> IoResult<crate::resp::Value> {
            let value = self.inputs.pop().expect("got more reads than expected");
            if value.eq_ignore_ascii_case("end") {
                return Err(super::IoError::EndOfInput);
            }
            Ok(value)
        }

        fn write_value(&mut self, value: crate::resp::Value) -> IoResult<usize> {
            let expected = self.outputs.pop().expect("got more writes than expected");
            assert_eq!(value, expected, "writes don't match");
            Ok(1)
        }
    }

    impl Drop for MockIo {
        fn drop(&mut self) {
            if std::thread::panicking() {
                return;
            }
            assert!(
                self.inputs.is_empty() || self.outputs.is_empty(),
                "got less reads and writes than expected: {self:?}"
            );
            assert!(
                self.inputs.is_empty(),
                "got less reads than expected: {:?}",
                self.inputs
            );
            assert!(
                self.outputs.is_empty(),
                "got less writes than expected: {:?}",
                self.outputs
            );
        }
    }

    #[test]
    #[should_panic(expected = "got less reads than expected")]
    fn mock_panics_on_too_few_reads() {
        _ = MockIo::new([Value::SimpleString("_".into())], []);
    }
    #[test]
    #[should_panic(expected = "got more reads than expected")]
    fn mock_panics_on_too_many_reads() {
        let mut io = MockIo::new([], []);
        _ = io.read_value().unwrap();
    }
    #[test]
    fn mock_produces_expected_reads() {
        let inputs = [
            Value::SimpleString("a".into()),
            Value::BulkString("b".into()),
        ];
        let mut io = MockIo::new(inputs.clone(), []);
        let res = [io.read_value().unwrap(), io.read_value().unwrap()];
        assert_eq!(res, inputs);
    }
    #[test]
    #[should_panic(
        expected = "assertion `left == right` failed: writes don't match\n  left: SimpleString(\"wrong\")\n right: SimpleString(\"exp\")"
    )]
    fn mock_panics_on_write_not_matcing() {
        let mut io = MockIo::new([], [Value::SimpleString("exp".into())]);
        io.write_value(Value::SimpleString("wrong".into())).unwrap();
    }

    #[test]
    #[should_panic(expected = "got less writes than expected")]
    fn mock_panics_on_too_few_writes() {
        _ = MockIo::new([], [Value::SimpleString("_".into())]);
    }

    #[test]
    #[should_panic(expected = "got more writes than expected")]
    fn mock_panics_on_too_many_writes() {
        let mut io = MockIo::new([], []);
        io.write_value(Value::SimpleString("_".into())).unwrap();
    }

    #[test]
    #[should_panic(expected = "got less reads and writes than expected")]
    fn mock_panics_on_too_few_reads_and_writes() {
        _ = MockIo::new(
            [Value::SimpleString("_".into())],
            [Value::SimpleString("_".into())],
        );
    }
}
