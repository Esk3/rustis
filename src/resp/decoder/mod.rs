use std::io::{BufRead, Read};

pub struct Decoder<R> {
    r: std::io::BufReader<R>,
}

impl<'a> Decoder<std::io::Cursor<&'a [u8]>> {
    pub fn decode_from_buf(buf: &'a [u8]) -> Result<Token, ()> {
        Self::new(std::io::Cursor::new(buf)).decode_token()
    }
}

impl<R> Decoder<R>
where
    R: std::io::Read,
{
    pub fn new(r: R) -> Self {
        Self {
            r: std::io::BufReader::new(r),
        }
    }

    pub fn decode_token(&mut self) -> Result<Token, ()> {
        let header = self.decode_header().unwrap();
        Ok(match header {
            Identifier::SimpleString => Token::String(self.decode_simple_string().unwrap()),
            Identifier::BulkString(size) => match self.decode_bulk_string(size).unwrap() {
                Some(buf) => Token::Buf(buf),
                None => Token::NullBuf,
            },
            Identifier::Array(size) => {
                self.decode_array();
                todo!()
            }
        })
    }

    fn decode_identifier(&mut self) {
        let mut buf = [0; 1];
        self.r.read_exact(&mut buf);
        match buf {
            _ => todo!(),
        }
    }

    fn decode_header(&mut self) -> Result<Identifier, ()> {
        //self.decode_identifier();
        let mut buf = [0; 1];
        self.r.read_exact(&mut buf).unwrap();
        Ok(match buf[0] {
            b'+' => Identifier::SimpleString,
            b'$' => Identifier::BulkString(self.read_digits().unwrap()),
            b'*' => Identifier::Array(self.read_digits().unwrap()),
            _ => return Err(()),
        })
    }
    fn read_digits(&mut self) -> std::io::Result<i64> {
        let mut buf = String::new();
        self.r.read_line(&mut buf)?;
        Ok(buf.parse().unwrap())
    }

    fn decode_simple_string(&mut self) -> std::io::Result<String> {
        let mut buf = String::new();
        self.r.read_line(&mut buf)?;
        buf.truncate(buf.len() - 2);
        Ok(buf)
    }

    fn decode_bulk_string(&mut self, size: i64) -> std::io::Result<Option<Vec<u8>>> {
        if size == -1 {
            return Ok(None);
        }
        assert!(size >= 0);
        let mut buf = vec![0; size as usize];
        self.r.read_exact(&mut buf)?;
        Ok(Some(buf))
    }

    fn decode_array(&mut self) {}

    fn line_feed(&mut self) -> std::io::Result<()> {
        let mut buf = [0; 2];
        self.r.read_exact(&mut buf)?;
        assert_eq!(buf, *b"\r\n", "{} {}", buf[0] as char, buf[1] as char);
        Ok(())
    }
}

pub enum Identifier {
    SimpleString,
    BulkString(i64),
    Array(i64),
}

#[derive(Debug)]
pub enum Token {
    String(String),
    Buf(Vec<u8>),
    Array(Box<Token>),
    NullBuf,
    NullArray,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_string() {
        let x = Decoder::decode_from_buf(b"+hello world\r\nmorestuff").unwrap();
        insta::assert_debug_snapshot!(x);
    }
}
