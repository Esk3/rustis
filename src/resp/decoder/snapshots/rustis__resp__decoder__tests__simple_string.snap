---
source: src/resp/decoder/mod.rs
expression: x
---
String(
    "hello world",
)
